### PR TITLE
Ter 59 modify readme so it is suitable for os consumption

### PR DESCRIPTION
**Pull request by HasanDevOps2003**
A diagram has been created and added to the README so that the user can have a visual representation of the architecture created. 

A "Best Practices" section has been added to the README to describe the best practices applied to the infrastructure.

A "Pre-requisites" section has been added to the README to tell the user what will be needed to use the module.

A "Naming Convention" section has been added to the README to explain to the user what naming conventions has been used for Terraform resources and AWS resources.

"Contributions" section has been added to the README to provide the user information as to where they can contribute to this module.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # AWS Ubuntu Desktop VM Module
 
+## Diagram
```
Comment (FaizanRaza-InterWeave) [inline README.md:3]: https://www.whizlabs.com/labs/create-automatic-snapshot-of-ebs-volume-using-amazon-dlm

Review this on how to show DLM

File: README.md [hunk 2]
```diff
@@ -1,5 +1,9 @@
...
+
+![image](https://github.com/InterweaveCloud/terraform_aws_ubuntu_desktop_vm/blob/TER-59-modify-readme-so-it-is-suitable-for-os-consumption/Architecture%20Diagram.png?raw=true)
```
Comment (FaizanRaza-InterWeave) [inline README.md]: this path needs to be relative, you put a link to a github feature branch

File: README.md [hunk 3]
```diff
@@ -70,12 +74,49 @@ From here click `Begin`, `Next` and `Authorise`.
 
 Currently you must SSH into the instance to apply the command but we will provide a remote exec in a future release.
 
+## Style Guidelines
```
Comment (FaizanRaza-InterWeave) [inline README.md]: This is not important put it much much lower

File: README.md [hunk 4]
```diff
@@ -70,12 +74,49 @@ From here click `Begin`, `Next` and `Authorise`.
...
+
+#### Networking
+
+- This is a single layer architecture so only one availability zone and one subnet is created to host the virtual machine.
+- An internet gateway with the appropriate routes are created to enable that the virtual machine has full access to the internet.
+
+#### Security
+
+- Spot instance security group only allows [SSH](https://www.ucl.ac.uk/isd/what-ssh-and-how-do-i-use-it#:~:text=SSH%20or%20Secure%20Shell%20is,web%20pages)%20and%20share%20data.), [VNC](https://discover.realvnc.com/how-to-use-vnc-connect-remote-desktop-software#:~:text=VNC%20stands%20for%20Virtual%20Network,right%20in%20front%20of%20it.) and [RDP](https://www.techtarget.com/searchenterprisedesktop/definition/Remote-Desktop-Protocol-RDP#:~:text=What%20is%20remote%20desktop%20protocol,their%20physical%20work%20desktop%20computers.) inbound access from IP addresses specified by the user. This ensures that only specific IP addresses are allowed access. 
+- Spot instance security group allows full outbound access so that the instance has full access to the internet.
```
Comment (FaizanRaza-InterWeave) [inline README.md]: not really a feature

File: README.md [hunk 5]
```diff
@@ -70,12 +74,49 @@ From here click `Begin`, `Next` and `Authorise`.
...
+### Best Practices
+
+#### Networking
+
+- This is a single layer architecture so only one availability zone and one subnet is created to host the virtual machine.
+- An internet gateway with the appropriate routes are created to enable that the virtual machine has full access to the internet.
+
+#### Security
+
+- Spot instance security group only allows [SSH](https://www.ucl.ac.uk/isd/what-ssh-and-how-do-i-use-it#:~:text=SSH%20or%20Secure%20Shell%20is,web%20pages)%20and%20share%20data.), [VNC](https://discover.realvnc.com/how-to-use-vnc-connect-remote-desktop-software#:~:text=VNC%20stands%20for%20Virtual%20Network,right%20in%20front%20of%20it.) and [RDP](https://www.techtarget.com/searchenterprisedesktop/definition/Remote-Desktop-Protocol-RDP#:~:text=What%20is%20remote%20desktop%20protocol,their%20physical%20work%20desktop%20computers.) inbound access from IP addresses specified by the user. This ensures that only specific IP addresses are allowed access.
```
Comment (FaizanRaza-InterWeave) [inline README.md]: Only allows SSH, VNC and RDP access

remove "This ensures that only specific IP addresses are allowed access."

File: README.md [hunk 6]
```diff
@@ -70,12 +74,49 @@ From here click `Begin`, `Next` and `Authorise`.
...
+
+Throughout the module, the naming convention used is as follows:
+- For Terraform resource names, the names consist of lowercase characters and _ to represent spaces in the name.
+- For AWS resource names, the names consist of lowercase characters and - to represent spaces in the name, with the <a name="input_prefix"></a> [prefix](#input\_prefix) variable attached to the beginning of the name and the <a name="input_environment"></a> [environment](#input\_environment).
+
+## Features
+
+### Best Practices
+
+#### Networking
```
Comment (FaizanRaza-InterWeave) [inline README.md:81]: These are not features.....

The feature is:

"Creates a new independent VPC with a public subnet to remove dependency on pre-existing VPCs."

File: README.md (hunk 5 above)
Comment (FaizanRaza-InterWeave) [inline README.md]: The SSH link doesnt work look at it in a preview.

File: README.md [hunk 7]
```diff
@@ -70,12 +74,49 @@ From here click `Begin`, `Next` and `Authorise`.
...
+
+- This is a single layer architecture so only one availability zone and one subnet is created to host the virtual machine.
+- An internet gateway with the appropriate routes are created to enable that the virtual machine has full access to the internet.
+
+#### Security
+
+- Spot instance security group only allows [SSH](https://www.ucl.ac.uk/isd/what-ssh-and-how-do-i-use-it#:~:text=SSH%20or%20Secure%20Shell%20is,web%20pages)%20and%20share%20data.), [VNC](https://discover.realvnc.com/how-to-use-vnc-connect-remote-desktop-software#:~:text=VNC%20stands%20for%20Virtual%20Network,right%20in%20front%20of%20it.) and [RDP](https://www.techtarget.com/searchenterprisedesktop/definition/Remote-Desktop-Protocol-RDP#:~:text=What%20is%20remote%20desktop%20protocol,their%20physical%20work%20desktop%20computers.) inbound access from IP addresses specified by the user. This ensures that only specific IP addresses are allowed access. 
+- Spot instance security group allows full outbound access so that the instance has full access to the internet.
+- A key pair is generated to allow authentication when accessing the instance using port 22(SSH).
```
Comment (FaizanRaza-InterWeave) [inline README.md]: it isnt generated. They generate the key pair. 

A user-provided SSH key is used to authenticate access to the instance through SSH.

File: README.md [hunk 8]
```diff
@@ -70,12 +74,49 @@ From here click `Begin`, `Next` and `Authorise`.
...
+
+#### Instance Configuration
+
+- An elastic IP address is allows you to remap your IP address onto another instance in the event that the spot instance is taken down, making the process of setting up the new instance much easier.
```
Comment (FaizanRaza-InterWeave) [inline README.md]: No not at all. 

The instance is never "taken down" i.e it is never terminated. The settings put the behavior as stop not terminate. 

Normally, when an EC2 instance is stopped and started, it is reallocated an IP address. An EIP ensures that it keeps the same IP address between stops and starts. 

"An elastic IP address is provisioned to preserve the public IP address of the instance between stops and starts. Note: The elastic IP address does not incur charges while the instance is running but will incur charges when stopped."

File: README.md [hunk 9]
```diff
@@ -70,12 +74,49 @@ From here click `Begin`, `Next` and `Authorise`.
...
+
+#### Instance Configuration
+
+- An elastic IP address is allows you to remap your IP address onto another instance in the event that the spot instance is taken down, making the process of setting up the new instance much easier. 
+
+#### EBS Backup
+
+- A Data Lifecycle Manager lifecycle policy is created which backups the instance daily before midnight and snapshots are retained for 1 week. This is ensures that in the event that the spot instance is taken down, you can use the snapshots of the EBS to recover your files.
+- An IAM role and policy is created which allows the DLM lifecycle policy to create and delete snapshots. It also allows DLM read access to instances,volumes and snapshots on EC2.
+- Snapshots are taken by the Data Lifecycle Manager and are stored in an S3 bucket managed by DLM.
```
Comment (FaizanRaza-InterWeave) [inline README.md]: remove s3 reference, its a leaky abstraction. We do not even need to know this, DLM abstracts this away from us, forget the module user

File: README.md [hunk 10]
```diff
@@ -70,12 +74,49 @@ From here click `Begin`, `Next` and `Authorise`.
...
+- A key pair is generated to allow authentication when accessing the instance using port 22(SSH).
+
+#### Instance Configuration
+
+- An elastic IP address is allows you to remap your IP address onto another instance in the event that the spot instance is taken down, making the process of setting up the new instance much easier. 
+
+#### EBS Backup
+
+- A Data Lifecycle Manager lifecycle policy is created which backups the instance daily before midnight and snapshots are retained for 1 week. This is ensures that in the event that the spot instance is taken down, you can use the snapshots of the EBS to recover your files.
+- An IAM role and policy is created which allows the DLM lifecycle policy to create and delete snapshots. It also allows DLM read access to instances,volumes and snapshots on EC2.
```
Comment (FaizanRaza-InterWeave) [inline README.md]: leaky abstraction, remove

File: README.md [hunk 11]
```diff
@@ -70,12 +74,49 @@ From here click `Begin`, `Next` and `Authorise`.
...
+- Spot instance security group allows full outbound access so that the instance has full access to the internet.
+- A key pair is generated to allow authentication when accessing the instance using port 22(SSH).
+
+#### Instance Configuration
+
+- An elastic IP address is allows you to remap your IP address onto another instance in the event that the spot instance is taken down, making the process of setting up the new instance much easier. 
+
+#### EBS Backup
+
+- A Data Lifecycle Manager lifecycle policy is created which backups the instance daily before midnight and snapshots are retained for 1 week. This is ensures that in the event that the spot instance is taken down, you can use the snapshots of the EBS to recover your files.
```
Comment (FaizanRaza-InterWeave) [inline README.md]: Taken down is not a technical term. Never use it again. 

"In the event of catastrophic failure or loss of data, these snapshots can be used to restore EC2 instances. Instances to be managed by this policy are identified by the following tag: `{AutomatedEbsBackups = true}`.

File: README.md [hunk 12]
```diff
@@ -70,12 +74,49 @@ From here click `Begin`, `Next` and `Authorise`.
...
+
 ## Resource Material
 
 [Ubuntu Blog on Ubuntu Desktop on GCP](https://ubuntu.com/blog/launch-ubuntu-desktop-on-google-cloud)  
 [Guide to install KDE Plasma 5](https://www.tecmint.com/install-kde-plasma-5-in-linux/)
 [Guide to generate SSH keys](https://docs.oracle.com/en/cloud/cloud-at-customer/occ-get-started/generate-ssh-key-pair.html)
 
+## Pre-requisites
+
+To be able to use this module you will need to have [AWS CLI](https://aws.amazon.com/cli/) installed on your local machine. An installation guide can be found [here](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html).
```
Comment (FaizanRaza-InterWeave) [inline README.md]: Why would you need AWS CLI?

Comment (FaizanRaza-InterWeave) [reply README.md]: You dont need AWS CLI

File: README.md [hunk 13]
```diff
@@ -168,4 +209,19 @@ module "ubuntu_desktop" {
...
+
+### Bug Reports & Feature Requests
+
+Please use the [issue tracker](https://github.com/InterweaveCloud/terraform_aws_ubuntu_desktop_vm/issues) to report any bugs or file feature requests.
+
+## Contributors
+| Name | Role |
+|------|------|
+| [Faizan Raza](https://www.linkedin.com/in/faizan-raza-997808206/) | Lead Developer |
+| [Muhammad Hasan](https://www.linkedin.com/in/muhammad-hasan-b2553221a/) | Developer |
```
Comment (FaizanRaza-InterWeave) [inline README.md]: you are the lead since you wrote it mainly

**Review by FaizanRaza-InterWeave — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (HasanDevOps2003) [reply README.md]: done

File: README.md (hunk 3 above)
Comment (HasanDevOps2003) [reply README.md]: done

File: README.md (hunk 4 above)
Comment (HasanDevOps2003) [reply README.md]: done

File: README.md (hunk 5 above)
Comment (HasanDevOps2003) [reply README.md]: done

File: README.md (hunk 6 above)
Comment (HasanDevOps2003) [reply README.md:81]: done

File: README.md (hunk 5 above)
Comment (HasanDevOps2003) [reply README.md]: done

File: README.md (hunk 7 above)
Comment (HasanDevOps2003) [reply README.md]: done

File: README.md (hunk 8 above)
Comment (HasanDevOps2003) [reply README.md]: done

File: README.md (hunk 9 above)
Comment (HasanDevOps2003) [reply README.md]: done

File: README.md (hunk 10 above)
Comment (HasanDevOps2003) [reply README.md]: done

File: README.md (hunk 11 above)
Comment (HasanDevOps2003) [reply README.md]: done

File: README.md (hunk 12 above)
Comment (HasanDevOps2003) [reply README.md]: done

File: README.md (hunk 13 above)
Comment (HasanDevOps2003) [reply README.md]: done

File: README.md (hunk 1 above)
Comment (HasanDevOps2003) [reply README.md:3]: DONE

File: README.md [hunk 14]
```diff
@@ -168,4 +194,29 @@ module "ubuntu_desktop" {
...
+
+Please use the [issue tracker](https://github.com/InterweaveCloud/terraform_aws_ubuntu_desktop_vm/issues) to report any bugs or file feature requests.
+
+## Contributors
+| Name | Role |
+|------|------|
+| [Muhammad Hasan](https://www.linkedin.com/in/muhammad-hasan-b2553221a/) | Lead Developer |
+| [Faizan Raza](https://www.linkedin.com/in/faizan-raza-997808206/) | Developer |
+
+## Style Guidelines
```
Comment (FaizanRaza-InterWeave) [inline README.md:211]: Its to do with contributing, so should be a level 2 title under contributing

File: README.md [hunk 15]
```diff
@@ -70,6 +74,28 @@ From here click `Begin`, `Next` and `Authorise`.
...
+- A key pair is created using a public key provided by the user to allow authentication when accessing the instance using port 22(SSH).
+
+#### Instance Configuration
+
+- An elastic IP address is provisioned to preserve the public IP address of the instance between stops and starts. Note: The elastic IP address does not incur charges while the instance is running but will incur charges when stopped.
+
+#### EBS Backup
+
+- A Data Lifecycle Manager lifecycle policy is created which backups the instance daily before midnight and snapshots are retained for 1 week. In the event of catastrophic failure or loss of data, these snapshots can be used to restore EC2 instances. Instances to be managed by this policy are identified by the following tag: {AutomatedEbsBackups = true}.
```
Comment (FaizanRaza-InterWeave) [inline README.md]: backups is a plural.

backs up.

And the tag needs to be wrapped between two `

**Review by FaizanRaza-InterWeave — APPROVED** [review]
resolve the comments yourself before merging. They are small so fine for you to do yourself

**Pull request merged**